src/main/java/org/folio/innreach/ccheduler/InitialContributionJobScheduler.java

```diff
@@ -0,0 +1,62 @@
+package org.folio.innreach.ccheduler;
+
+import com.google.common.cache.Cache;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.innreach.batch.contribution.service.InitialContributionEventProcessor;
+import org.folio.innreach.domain.entity.TenantInfo;
+import org.folio.innreach.domain.service.impl.TenantScopedExecutionService;
+import org.folio.innreach.repository.JobExecutionStatusRepository;
+import org.folio.innreach.repository.TenantInfoRepository;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+public class InitialContributionJobScheduler {
+  private final TenantScopedExecutionService executionService;
+  private final JobExecutionStatusRepository jobExecutionStatusRepository;
+  private final InitialContributionEventProcessor eventProcessor;
+  private final TenantInfoRepository tenantRepository;
+  @Value(value = "${initial-contribution.fetch-limit}")
+  private int recordLimit;
+  @Value(value = "${initial-contribution.item-pause}")
+  private int itemPause;
+  private final Cache<String, List<String>> tenantDetailsCache;
```

Comment:
I think we need to add an `AtomicInteger` property called `currentRequestCount` or similar here which can be used to track the count of contribution requests that are in the state of being in progress. This can be incremented on start and de-incremented on success or failure. This integer can be passed as an argument to the `@Async` method when it is called each time.

This will allow us to set the scheduler interval to be very low, like say 100 milliseconds. Each time it fires we first check the the `currentRequestCount` and compare it to config value we already have for the fetch limit and if there are no "slots" available (`fetchLimit == currentRequestCount`, return and wait for the next interval to pass. If a slot is available, increment the counter and start the request.

If we don't have something like this here is what I think will happen:
* If the interval is too brief the OS thread scheduler will spawn a new thread each time it fires, which will eventually cause OOM unless we somehow get the interval "just right".
* If the interval is too long, we may not get OOM (at least not right away), but performance will suffer since it won't always be doing work processing requests.

Even in the case when the interval is set to a "long time", we could still get ourselves into the first situation if inn reach happens to be performing really slow for a period of time. Lots of requests will get queued up each time it fires, and if inn-reach is really really slow, the same OOM situation will eventually occur.

Anyway I think this is an easy problem to fix. See what you think.

cc @kalibek-epam

Reply:
@steveellis the proper interval and amount of records per single invocation is the key part here. They are configurable parameters and their goal is to allow flexibility to find values to avoid the OOMs. On the other hand, what you propose is effectively introducing state inside stateless service. The goal of the outbox approach was to avoid it. Whatever state we may need like `currentRequestCount` should be persisted in the database.

Reply:
@kalibek-epam We could use the database to get the `currentRequestCount` instead of the `AtomicInteger` bound to the service. I know that we have the in progress state represented there in the db.

But I think getting this from the db makes the app harder to reason about. Consider the following.

If we have 2 containers running our app, container A and B, and each container has a CPU capacity of 4 vCPUs. Fetch the `currentRequestCount` from the db.  At any given time the `currentRequestCount` will be <= 8, the total count across all app instances. Now let's say container A happens to have one of its requests taking a really long time. Will the `currentRequestCount` if obtained from the database represent this state?

To put it another way: how does the container instance translate this count to its local state, which is what we need to know. Will the `fetchLimit` be equal to the sum of CPU capacity of all containers? It could be set to that. And if so container A would see that it doesn't have capacity to queue the next request, but so would container B. Also each time we add a container to the cluster of containers, we need to update this configurable `fetchLimit`.

How could this go wrong? Lets imagine one or two long running requests lasts 20 seconds, and our interval is firing every 100 milliseconds. If the container doesn't have real awareness of its own requests, this could mean that the scheduler would queue 20 * 10 requests prematurely in container A, which puts us back in the same situation we're trying to avoid.

I think all of these problems go away if we just track the container's own request count inside of the container. The most useful count here is the count as it relates to each container's cpu capacity.

Reply:
I think we agreed today to try it with the db-oriented approach. Let's see how it goes.